kore/src/Kore/Attribute/Axiom/Concrete.hs

```diff
@@ -63,27 +74,42 @@ concreteSymbol =
         }
 
 -- | Kore pattern representing the @concrete@ attribute.
-concreteAttribute :: AttributePattern
-concreteAttribute = attributePattern_ concreteSymbol
+concreteAttribute :: [UnifiedVariable Variable] -> AttributePattern
+concreteAttribute = attributePattern concreteSymbol . map attributeVariable
 
 parseConcreteAttribute
-    :: forall variable
-    .  Ord variable
-    => FreeVariables variable
+    :: FreeVariables Variable
     -> AttributePattern
-    -> Concrete variable
-    -> Parser (Concrete variable)
+    -> Concrete Variable
+    -> Parser (Concrete Variable)
 parseConcreteAttribute freeVariables =
-    parseBoolAttributeAux iso concreteId
+    Parser.withApplication concreteId parseApplication
   where
-    iso :: Lens.Iso' (Concrete variable) Bool
-    iso =
-        Lens.iso
-            isConcrete
-            (\b -> Concrete $ if b then freeVariables else mempty)
-
-instance From (Concrete variable) Attributes where
-    from = toBoolAttributesAux (Lens.to isConcrete) concreteAttribute
+    parseApplication
+        :: [Sort]
+        -> [AttributePattern]
+        -> Concrete Variable
+        -> Parser (Concrete Variable)
+    parseApplication params args (Concrete concreteVars) = do
+        Parser.getZeroParams params
+        vars <- mapM getVariable args
+        mapM_ checkFree vars
+        let newVars =
+                if null vars then freeVariables else foldMap freeVariable vars
+        mapM_ (checkDups concreteVars) (getFreeVariables newVars)
```

Comment:
Could you add a comment explaining why it is necessary to check for duplicates even if `vars` is empty? I had to read this a few times before it was clear to me.